test/suite/functional/Generator/SchemaMapper/complexOneOf.yaml

```diff
@@ -0,0 +1,85 @@
+openapi: 3.0.0
+info:
+  title: Complex OneOf with Nested Structures
+  version: 1.0.0
+
+paths:
+  /example:
+    get:
+      responses:
+        '200':
+          description: An example of a complex oneOf structure
+          content:
+            application/json:
+              schema:
+                oneOf:
+                  - $ref: '#/components/schemas/Animal'
+                  - $ref: '#/components/schemas/Machine'
+                discriminator:
```

Comment:
even though we're adding support for oneOf we're only trying to build the objects based on the response and completely ignoring this "discriminator" key...

I think ideally we should add support for it as well, if discriminator key is present we won't try/catch until we find the right schema, instead we'll build directly the right schema based on the "mapping" when that key exists or on "propertyName" directly (as described in https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/ under "Discriminator" section).

That or we check if "discriminator" key is present we warn that that feature is not support like we have right now with oneOf... or else we might imply we support that but we don't, and there's an edge case where two oneOf schemas can have same required properties but different optional properties and if only required properties are present in response it won't really know which one to build if we don't check "discriminator"...

Reply:
If there are two or more oneOf schemas with same required properties but different optional properties: may be in this case we should detect it and build "the least common denominator" parent schema object?

```yaml
oneOf:
  - type: object
    required:
      - name
    properties:
      name:
        type: string
      color:
        type: string
        enum: [blue, red]
  - type: object
    required:
      - name
    properties:
      name:
        type: string
      size:
        type: string
        enum: [small, big]
```

```php
class Named {
  public string $name;
}

class Colored extends Named {
  public string $color;
}

class Sized extends Named {
  public string $size;
}
```

Ideally user would define the schema using inheritance, then it will be much easier to detect.

Reply:
We already support allOf which could be used to model something like this (Colored is allOf Named + color, Sized is allOf Named + size etc.)

but I think that's not the case with oneOf with your example, just because they share the same required properties doesn't mean they are "something in common", I don't think we should try to identify these similarities and group them in a separate object or smth

I like how you did each oneOf schema is a separate object with its own properties etc., just saying that we should also support "discriminator" because in these cases where oneOf schemas are similar we won't know for sure which one to build without the "hint" from the discriminator.... that OR throw warning when "discriminator" is present saying we don't support that feature yet...

Reply:
I agree, the best is to fail the generation if there's no discriminator.

Reply:
we can print a warning saying it's not supported still (like we do currently with the oneOf), instead of failing altogether, and ofc add support for it later

it's possible to try recognizing the right schema without "discriminator" (after all it's an optional property so we should be able to generate client with oneOf without discriminator) and this PR is doing exactly that.

my point is more like to highlight to people that the tool might not work properly if discriminator is used in the schema